enhancements/installer/ipi-publishing-status.md

```diff
@@ -0,0 +1,199 @@
+---
+title: expose-cluster-publishing-status
+authors:
+  - "@flavianmissi"
+reviewers: # Include a comment about what domain expertise a reviewer is expected to bring and what area of the enhancement you expect them to focus on. For example: - "@networkguru, for networking aspects, please look at IP bootstrapping aspect"
+  - "@patrickdillon, installer"
+  - "@Miciah, edge networking"
+approvers: # A single approver is preferred, the role of the approver is to raise important questions, help ensure the enhancement receives reviews from all applicable areas/SMEs, and determine when consensus is achieved such that the EP can move forward to implementation.  Having multiple approvers makes it difficult to determine who is responsible for the actual approval.
+  - TBD
+api-approvers: # In case of new or modified APIs or API extensions (CRDs, aggregated apiservers, webhooks, finalizers). If there is no API change, use "None"
+  - "@bparees"
+creation-date: yyyy-mm-dd
+last-updated: yyyy-mm-dd
+tracking-link: # link to the tracking ticket (for example: Jira Feature or Epic ticket) that corresponds to this enhancement
+  - [IR-302](https://issues.redhat.com/browse/IR-302)
+---
+
+# Expose Cluster Publishing Status
+
+## Summary
+
+IPI clusters can be configured to be public or private. This is done in the
+install-config.yaml using the `Publish` parameter. There is no current standard
+defined for how components should access this parameter to bootstrap themselves
+as public (External) or private (Internal).
+
+## Motivation
+
+A component should be able to know how to bootstrap itself according to user
+preference. A central place providing the cluster publishing status at install
+time makes more sense than the installer having to individually configure
+various components.
+
+### User Stories
+
+* As the author of the image registry operator, I want to easily discover the
+cluster publishing status, so that I can configure access the cloud storage objects
+accordingly.
+
+### Goals
+
+Expose the cluster publishing status used at install time in the Infrastructure
+object.
+
+### Non-Goals
+
+This enhancement does not cover support for updating the cluster publishing
+status on running clusters. This property is meant to reflect the `publish`
+parameter as it was set in the install-config.yaml.
+
+## Proposal
+
+The Infrastructure API is extended, adding the `Publish` field to the
+`InfrastructureStatus`.
+
+```go
+type InfrastructureStatus struct {
+	// ...
+
+	// publish indicates whether the cluster was installed as public or private.
+	// The default is 'External', which means the cluster will be publicly
+    // available on the internet.
+	// The 'Internal' mode indicates a private cluster, and operators should
+    // configure operands accordingly.
+    // +kubebuilder:default=External
+    // +kubebuilder:validation:Enum=Internal;External
```

Comment:
ptal @JoelSpeed

Reply:
To apply a default, you need to have `omitempty` and make the field optional with `// +optional`.

That said, what happens on upgrades here? If the field isn't already set you'll set it by default to External? That doesn't seem right. I think you need to add this with an empty default to support the upgrade case unless you can reliably correct this.

Is the field applicable to all platforms or only some platforms?

Reply:
You are correct, we need the default to be empty - gave that a shot but unsure about syntax, PTAL.

> Is the field applicable to all platforms or only some platforms?

Good question. While the `publish` option is widely available among supported platforms, I don't think it's available for _all_ of them. Perhaps the empty value is good enough for the ones where this does not apply?

Reply:
Typically where a field does not apply to all platforms, we put the field within the platformSpec or platformStatus discriminated unions. Of the IPI platforms we support, can you list which do and which do not support the publish option? That enumeration is probably good to include in the EP anyway btw so we know which platforms are affected by this change

Reply:
According to the [installer source code](https://github.com/openshift/installer/blob/master/pkg/types/validation/installconfig.go#L150-L156), changing the `publish` field in the installer-config.yaml to `Internal` is _supported_ on the following:
* AWS
* Azure
* GCP
* Alibaba
* IBM Cloud
* PowerVS

I think it would make sense to keep the `Publish` field platform specific as per your suggestion - changed the code, PTAL.